docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -110,6 +111,52 @@ We need to be able to process transactions and roll-back state updates if a tran
 
 We identified use-cases, where modules will need to save an object commitment without storing an object itself. Sometimes clients are receiving complex objects, and they have no way to prove a correctness of that object without knowing the storage layout. For those use cases it would be easier to commit to the object without storing it directly.
 
+
+### Remove MultiStore
+
+IAVL based store adds an additional layer in the SDK store construction - the `MultiStore` structure. The multistore exists to support the modularity of the Cosmos SDK - each module is using its own instance of IAVL, but in the current implementation, all instances share the same database.
+The latter indicates, however, that the implementation doesn't provide true modularity. Instead it causes problems related to race condition and sync problems (eg: [\#6370](https://github.com/cosmos/cosmos-sdk/issues/6370)).
+
+We propose to remove the multistore concept from the SDK, and to use a single instance of `SC` and `SS` in a `rootStore` object. To avoid confusion, we should rename the `MultiStore` interface to `RootStore` and make sure that the `rootStore` object implements `RootStore`.
+
+Moreover, to improve usability, we should extend the `KVStore` interface with _prefix store_. This will allow module developers to bind a store to a namespace for module sub-components:
+
+```go
+type KVStore interface {
+    ... current KVStore
+
+    WithPrefix(prefix string) KVStore
```

Comment:
I agree, mentioned this above https://github.com/cosmos/cosmos-sdk/pull/9355#issuecomment-874502492

Reply:
It makes sense for this API to exist on the `RootStore` as that is how I imagine module's will request access to a prefixed store. That abstraction makes sense to me.

Reply:
If this `WithPrefix` method just takes a string and compresses it, it shouldn't be used by modules for prefix stores

Reply:
Then again maybe that's something we do want modules to have to manage key prefixes 🤔

Reply:
I think it makes sense to have it in `KVStore` - this way we can encapsulate context and implement optimization inside.

Reply:
In SDK, modules usually have _key families_ - objects with a common prefix. For  example, in `x/bank` we have: supply, DenomMetadata, DenomAddress, Balances. In one of the architecture updates to the module we want that the module will receive it's own store. In x/bank, today we have:
```
ctx.KVStore(storeKey)  // where storeKey = "bank"
```
This doesn't follow OCAPS, because in x/bank I coud do `ctx.KVStore("gov")`. Instead we want:
```
bankStore := ctx.KVStore(goCtx)
// or
bankStore := accessor.KVStore(goCtx)
```
Then I can write:
```
balanceStore := bankStore.WithPrefix("BalancesPrefix")
```
and us it in some DAL responsible for balance management.

Reply:
If `WithPrefix` is confusing then we can rename it to `Substore` or `Subnamespace`.

Reply:
But what advantage does this have over:
`balanceStore := PrefixStore(bankStore, "BalancesPrefix")`

What kind of optimization would warrant adding this to the KVStore interface?

Reply:
@roysc , you are right, `PrefixStore` function is better than extending the interface.